In [167]:
import pandas as pd

### TODO:
- [x] Daten einlesen
- [x] Daten angucken
- [x] Daten bereinigen
- [x] Daten zusammenführen
- [x] Daten speichern

## Daten einlesen

In [168]:
path_to_kunden = "./data/kunden.csv"
path_to_geo = "./data/geo.txt"
path_to_besucher = "./data/besucher.csv"

In [169]:
kunden_df = pd.read_csv(path_to_kunden, delimiter=";")
besucher_df = pd.read_csv(path_to_besucher, delimiter=";")
geo_df = pd.read_csv(path_to_geo, delimiter="\t")

## Daten angucken
In diesem Abschnitt gucken wir uns jeden einzelnen Dataframe an, um festzustellen wo möglicherweise einzelne Datenpunkte fehlen (z.B. None) und Daten angeglichen werden müssen.

### kunden_df angucken

In [170]:
kunden_df

Alter  Einkommen      Preis  Geschlecht  Zeit KundeNr
0        64    66894.0   88160.31         1.0  43.0   K0310
1        54    77644.0  103145.70         1.0  40.0   K1042
2        55    44341.0   80565.16         0.0  37.0   K0382
3        49    67271.0   83949.89         0.0  42.0   K0498
4        46    49832.0   93781.58         0.0  41.0   K0552
...     ...        ...        ...         ...   ...     ...
1099     70    69581.0  103751.58         1.0  51.0   K0467
1100     38    48301.0   64230.98         1.0  33.0   K0122
1101     67    63547.0  103123.57         1.0  41.0   K1045
1102     37    32686.0   49215.87         1.0  36.0   K1096
1103     55    71240.0  106132.59         1.0  49.0   K0861

[1104 rows x 6 columns]

In [171]:
kunden_df.dtypes

Alter           int64
Einkommen     float64
Preis         float64
Geschlecht    float64
Zeit          float64
KundeNr        object
dtype: object

In [172]:
kunden_df.isnull().sum()

Alter         0
Einkommen     0
Preis         0
Geschlecht    5
Zeit          0
KundeNr       0
dtype: int64

In [173]:
kunden_df.describe()

Alter     Einkommen          Preis   Geschlecht         Zeit
count  1104.000000  1.104000e+03    1104.000000  1099.000000  1104.000000
mean     50.481884  9.644748e+05   88975.198714     0.637853    44.375906
std      12.225059  3.009470e+07   24464.067781     0.480840    10.848040
min      30.000000 -6.284037e+04   21471.650000     0.000000    12.000000
25%      40.000000  4.663850e+04   70609.422500     0.000000    37.000000
50%      51.000000  5.901300e+04   88931.455000     1.000000    44.000000
75%      61.000000  7.029800e+04  107093.375000     1.000000    52.000000
max     120.000000  1.000000e+09  165482.310000     1.000000    80.000000

In [174]:
kunden_df.loc[kunden_df['Einkommen'] < 0]

Alter     Einkommen     Preis  Geschlecht  Zeit KundeNr
1082     59 -62840.374523  92837.59         0.0  46.0   K1083

In [175]:
kunden_df.loc[kunden_df['Einkommen'] > 1000000]

Alter     Einkommen     Preis  Geschlecht  Zeit KundeNr
408     68  1.000000e+09  87209.72         0.0  52.0   K0701

In [176]:
kunden_df.loc[kunden_df['Alter'] > 70]

Alter  Einkommen      Preis  Geschlecht  Zeit KundeNr
142    100    88749.0  119134.23         0.0  64.0   K0494
711    120    61857.0  104214.07         1.0  52.0   K0039

### besucher_df angucken

In [177]:
besucher_df

Alter           Einkommen  Geschlecht  Zeit KundeNr
0       30    39919,0953348874           0  29,0   K1814
1       41   53430,67308587748           0  27,0   K1544
2       64    73163,2169267002           0  32,0   K1945
3       41  36761,482581315344           0  44,0   K1825
4       35   42684,33163178136           1  33,0   K1144
..     ...                 ...         ...   ...     ...
886     32   34391,82496545645           1  48,0   K1211
887     41   68516,49543941053           0  34,0   K1375
888     65   89685,33344089528           0  23,0   K1965
889     68   95714,30526812703           0  32,0   K1540
890     49    73319,1870845747           0  27,0   K1207

[891 rows x 5 columns]

In [178]:
besucher_df.dtypes

Alter          int64
Einkommen     object
Geschlecht     int64
Zeit          object
KundeNr       object
dtype: object

In [179]:
besucher_df.isnull().sum()

Alter         0
Einkommen     0
Geschlecht    0
Zeit          0
KundeNr       0
dtype: int64

In [180]:
# durch falsche Datentypen hier wenig sinnvoll: muss nochmal wiederholt werden nach Bereinigung
besucher_df.describe()

Alter  Geschlecht
count  891.000000  891.000000
mean    49.881033    0.578002
std     11.865889    0.494156
min     30.000000    0.000000
25%     40.000000    0.000000
50%     49.000000    1.000000
75%     60.000000    1.000000
max     70.000000    1.000000

### geo_df angucken

In [181]:
geo_df

KundeNr          Niederlassung
0      K0001                Sachsen
1      K0002                    NRW
2      K0003                 Hessen
3      K0004                 Bayern
4      K0005                 Bayern
...      ...                    ...
1990   K1991                 Hessen
1991   K1992      Baden-Württemberg
1992   K1993    Nordrhein-Westfalen
1993   K1994                 Berlin
1994   K1995  Berlin-Charlottenburg

[1995 rows x 2 columns]

In [182]:
geo_df.dtypes

KundeNr          object
Niederlassung    object
dtype: object

In [183]:
geo_df.isnull().sum()

KundeNr          0
Niederlassung    0
dtype: int64

In [184]:
geo_df['Niederlassung'].value_counts()

Niederlassung
Baden-Württemberg        382
Bayern                   338
Nordrhein-Westfalen      322
Berlin                   230
Niedersachsen            169
Hessen                   165
Sachsen                  162
Thüringen                 89
Hamburg                   67
Brandenburg               60
NRW                        3
Düsseldorf                 2
Berlin-Charlottenburg      2
Berlin-Mitte               2
BERLIN                     2
Name: count, dtype: int64

## Daten bereinigen
Folgende Probleme wurden im vorherigen Abschnitt festgestellt:
- kunden_df enthält 5 Reihen, wo das Geschlecht nicht ausgefüllt ist (isnull())
- kunden_df enthält einen Reihe, wo das Gehalt negativ ist. Es wird davon ausgegangen, dass es sich hierbei um einen Tippfehler handelt
- kunden_df enthält Ausreißer
  - Alter: 100 und 120 sind nicht realitische Alter für Autokäufer. Die nächste Person im Datensatz ist 70 Jahre alt. Die beiden Einträge werden durch den Modalwert ersetzt.
  - Einkommen: 1.000.000.000 (1 Mrd) und Einkommen 0 sind unrealistische Jahresgehälter, weshalb es durch den Median ersetzt wird (Modalwert nicht möglich, da alle Einkommensangaben einzigartig sind).
- Die Spalten in besucher_df haben nicht sinnvoll gewählte Datentypen (dtypes). Diese müssen angeglichen werden, sodass es später beim Zusammenführen mit kunden_df nicht zu Problemen kommt.
  - Einkommen: object -> float64
  - Geschlecht: int64 -> float64
  - Zeit: object -> float64
- geo_df enthält Unstimmigkeiten in der Spalte Niederlassung:
  - NRW, Nordrhein-Westfalen und Düsseldorf beschreiben das gleiche Bundesland, sind aber unterschiedlich angegeben
  - Berlin, BERLIN, Berlin-Charlottenburg und Berlin-Mitte beschreiben ebenfalls das gleiche Bundesland


### kunden_df bereinigen

Problem fehlende Einträge Geschlecht:

Wir wollen die Reihen nicht löschen, weil uns dann wichtige Kundeninformationen wie Einkommen und Alter verloren gehen. Weil der überwiegende Teil der Kunden (ca. 700) männlich ist, nehmen wir die fehlenden Einträge ebenfalls als männlich an.

Begründung:
Wählt man eine zufällige Person aus dem Datensatz aus, so ist die Wahrscheinlichkeit, dass diese männlich ist deutlich höher (ca. 64%), als dass diese weiblich ist (ca. 36%).

In [185]:
cleaned_kunden_df = kunden_df.fillna(1.0)

In [186]:
cleaned_kunden_df.isnull().sum()

Alter         0
Einkommen     0
Preis         0
Geschlecht    0
Zeit          0
KundeNr       0
dtype: int64

Negativer Eintrag df_kunden Einkommen: 

In [187]:
cleaned_kunden_df.loc[cleaned_kunden_df['Einkommen'] < 0, 'Einkommen'] *= -1

In [188]:
cleaned_kunden_df[cleaned_kunden_df['Einkommen'] < 0]

Empty DataFrame
Columns: [Alter, Einkommen, Preis, Geschlecht, Zeit, KundeNr]
Index: []

Modalwert für Alter bestimmen:
    

In [189]:
age_mode = cleaned_kunden_df.loc[cleaned_kunden_df['Alter'] < 70, 'Alter'].mode()[0]

Ausreißer Alter 100 und 120 zu Modalwert ändern:

In [190]:
cleaned_kunden_df.loc[cleaned_kunden_df['Alter'] > 70, 'Alter'] = age_mode

In [191]:
cleaned_kunden_df[cleaned_kunden_df['Alter'] > 70]

Empty DataFrame
Columns: [Alter, Einkommen, Preis, Geschlecht, Zeit, KundeNr]
Index: []

Median der Einkommen ohne Ausreißer bestimmen:

In [192]:
income_median = cleaned_kunden_df.loc[(cleaned_kunden_df['Einkommen'] < 1000000) & (cleaned_kunden_df['Einkommen'] != 0), 'Einkommen'].median()

Ausreißer (1 mrd) wird zum Median geändert, da nächstgelegner Wert 100.000 ist und 1 Mrd. Einkommen unrealistisch ist:  

In [193]:
cleaned_kunden_df.loc[cleaned_kunden_df['Einkommen'] > 1000000, 'Einkommen'] = income_median

In [194]:
cleaned_kunden_df[cleaned_kunden_df['Einkommen'] > 1000000]

Empty DataFrame
Columns: [Alter, Einkommen, Preis, Geschlecht, Zeit, KundeNr]
Index: []

In [195]:
cleaned_kunden_df[cleaned_kunden_df["Einkommen"] == 0]

Alter  Einkommen      Preis  Geschlecht  Zeit KundeNr
966     69        0.0  125140.05         1.0  65.0   K1021

Ausreißer (0 Einkommen) wird zum Median geändert, da nächstgelegener Wert 14547 ist und Einkommen von 0 unrealisitisch ist:

In [196]:
cleaned_kunden_df.loc[cleaned_kunden_df['Einkommen'] == 0, 'Einkommen'] = income_median

In [197]:
cleaned_kunden_df[cleaned_kunden_df["Einkommen"] == 0]

Empty DataFrame
Columns: [Alter, Einkommen, Preis, Geschlecht, Zeit, KundeNr]
Index: []

### besucher_df bereinigen

In [198]:
besucher_df.head()

Alter           Einkommen  Geschlecht  Zeit KundeNr
0     30    39919,0953348874           0  29,0   K1814
1     41   53430,67308587748           0  27,0   K1544
2     64    73163,2169267002           0  32,0   K1945
3     41  36761,482581315344           0  44,0   K1825
4     35   42684,33163178136           1  33,0   K1144

In [199]:
besucher_df.dtypes

Alter          int64
Einkommen     object
Geschlecht     int64
Zeit          object
KundeNr       object
dtype: object

In [200]:
cleaned_besucher_df = besucher_df.copy()

# Um Zeit und Einkommen zu float64 zu konvertieren, muss das Komma jeweils durch einen Punkt ersetzt werden
cleaned_besucher_df['Einkommen'] = cleaned_besucher_df['Einkommen'].str.replace(',', '.')
cleaned_besucher_df['Zeit'] = cleaned_besucher_df['Zeit'].str.replace(',', '.')

cleaned_besucher_df = cleaned_besucher_df.astype({'Geschlecht': 'float64', 'Einkommen': 'float64', 'Zeit': 'float64'})

In [201]:
cleaned_besucher_df.head()

Alter     Einkommen  Geschlecht  Zeit KundeNr
0     30  39919.095335         0.0  29.0   K1814
1     41  53430.673086         0.0  27.0   K1544
2     64  73163.216927         0.0  32.0   K1945
3     41  36761.482581         0.0  44.0   K1825
4     35  42684.331632         1.0  33.0   K1144

In [202]:
cleaned_besucher_df.dtypes

Alter           int64
Einkommen     float64
Geschlecht    float64
Zeit          float64
KundeNr        object
dtype: object

In [203]:
cleaned_besucher_df.describe()

Alter      Einkommen  Geschlecht        Zeit
count  891.000000     891.000000  891.000000  891.000000
mean    49.881033   59367.936188    0.578002   35.166105
std     11.865889   16164.358221    0.494156   10.476802
min     30.000000   10802.033307    0.000000    3.000000
25%     40.000000   47816.611417    0.000000   28.000000
50%     49.000000   59163.823681    1.000000   35.000000
75%     60.000000   71001.378839    1.000000   43.000000
max     70.000000  108569.087021    1.000000   68.000000

### geo_df bereinigen

In [204]:
cleaned_geo_df = geo_df.copy()
cleaned_geo_df.loc[cleaned_geo_df['Niederlassung'] == 'NRW', 'Niederlassung'] = 'Nordrhein-Westfalen'
cleaned_geo_df.loc[cleaned_geo_df['Niederlassung'] == 'Düsseldorf', 'Niederlassung'] = 'Nordrhein-Westfalen'
cleaned_geo_df.loc[cleaned_geo_df['Niederlassung'] == 'BERLIN', 'Niederlassung'] = 'Berlin'
cleaned_geo_df.loc[cleaned_geo_df['Niederlassung'].str.startswith('Berlin-'), 'Niederlassung'] = 'Berlin'

In [205]:
cleaned_geo_df['Niederlassung'].value_counts()

Niederlassung
Baden-Württemberg      382
Bayern                 338
Nordrhein-Westfalen    327
Berlin                 236
Niedersachsen          169
Hessen                 165
Sachsen                162
Thüringen               89
Hamburg                 67
Brandenburg             60
Name: count, dtype: int64

## Daten zusammenführen

![Visualisierung der Merge Typen in Pandas](https://miro.medium.com/v2/resize:fit:400/format:webp/1*9eH1_7VbTZPZd9jBiGIyNA.png)

Die Daten werden mithilfe eines Outer Joins zusammengeführt. Das bedeutet, dass wenn z.B. ein Eintrag in besucher_df existiert aber nicht in kunden_df, so wird der Eintrag bei der Zusammenführung zwar übernommen, aber die Spalten aus kunden_df bleiben zunächst leer (NaN). Diese NaN Werte werden im Anschluss durch 0 ersetzt.
Um in der späteren Bearbeitung der Aufgaben Kunden und Nicht-Käufer von einander zu trennen, kann auf folgende Weise gefiltert werden:
- Kunden: `combined_df.loc[combined_df['Preis'] != 0]`
- Nicht-Käufer (Besucher abzüglich Kunden): `combined_df.loc[combined_df['Preis'] == 0]`

In [206]:
combined_df = cleaned_kunden_df.copy()
combined_df = combined_df.merge(cleaned_besucher_df, how='outer', on=['KundeNr', 'Alter', 'Einkommen', 'Geschlecht', 'Zeit'])
combined_df = combined_df.merge(cleaned_geo_df, how='outer', on='KundeNr')
combined_df = combined_df.fillna(0)
combined_df

Alter     Einkommen      Preis  Geschlecht  Zeit KundeNr  \
0        64  66894.000000   88160.31         1.0  43.0   K0310   
1        54  77644.000000  103145.70         1.0  40.0   K1042   
2        55  44341.000000   80565.16         0.0  37.0   K0382   
3        49  67271.000000   83949.89         0.0  42.0   K0498   
4        46  49832.000000   93781.58         0.0  41.0   K0552   
...     ...           ...        ...         ...   ...     ...   
1990     32  34391.824965       0.00         1.0  48.0   K1211   
1991     41  68516.495439       0.00         0.0  34.0   K1375   
1992     65  89685.333441       0.00         0.0  23.0   K1965   
1993     68  95714.305268       0.00         0.0  32.0   K1540   
1994     49  73319.187085       0.00         0.0  27.0   K1207   

          Niederlassung  
0                Bayern  
1                Berlin  
2     Baden-Württemberg  
3                Bayern  
4                Hessen  
...                 ...  
1990             Hessen  
1991            Sachsen  
1992             Bayern  
1993             Bayern  
1994      Niedersachsen  

[1995 rows x 7 columns]

## Daten speichern
Die einzelnen Dataframes sowie der zusammengeführte Dataframe werden im Ordner `./cleaned_data` gespeichert.

In [207]:
combined_df.to_csv('./data/df_final.csv', index=False)
cleaned_kunden_df.to_csv('./data/cleaned_kunden.csv', index=False)
cleaned_besucher_df.to_csv('./data/cleaned_besucher.csv', index=False)
cleaned_geo_df.to_csv('./data/cleaned_geo.csv', index=False)